qiskit/circuit/add_control.py

```diff
@@ -149,15 +147,43 @@ def control(operation: Union[Gate, ControlledGate],
                             use_basis_gates=True)
             elif rule[0].name == 'u1':
                 qc.mcu1(rule[0].params[0], q_control, q_target[rule[1][0].index])
-            elif rule[0].name == 'cx':
-                qc.mct(q_control[:] + [q_target[rule[1][0].index]], q_target[rule[1][1].index],
+            elif rule[0].name == 'cx' or rule[0].name == 'ccx':
+                additional_control_bits = [bit.index for bit in rule[1][:-1]]
+                qc.mct(q_control[:] + q_target[additional_control_bits],
+                       q_target[rule[1][-1].index],
                        q_ancillae)
             elif rule[0].name == 'x':
-                qc.mcx(q_control, q_target[rule[1][0].index], q_ancillae)
+                qc.mct(q_control[:], q_target[rule[1][0].index], q_ancillae)
             elif rule[0].name == 'z':
-                qc.cz(q_control, q_target[rule[1][0].index])
+                from qiskit.circuit.library.standard_gates import ZGate
+                mcz = ZGate().control(num_ctrl_qubits)
+                qargs = q_control[:] + q_target[[bit.index for bit in rule[1]]]
+                qc.append(mcz, qargs)
             elif rule[0].name == 'y':
-                qc.cy(q_control, q_target[rule[1][0].index])
+                from qiskit.circuit.library.standard_gates import YGate
+                mcy = YGate().control(num_ctrl_qubits)
+                qargs = q_control[:] + q_target[[bit.index for bit in rule[1]]]
+                qc.append(mcy, qargs)
+            elif rule[0].name == 'h':
+                from qiskit.circuit.library.standard_gates import HGate
+                mch = HGate().control(num_ctrl_qubits)
+                qargs = q_control[:] + q_target[[bit.index for bit in rule[1]]]
+                qc.append(mch, qargs)
+            elif rule[0].name == 'rx':
+                from qiskit.circuit.library.standard_gates import RXGate
+                mcrx = RXGate(rule[0].params[0]).control(num_ctrl_qubits)
+                qargs = q_control[:] + q_target[[bit.index for bit in rule[1]]]
+                qc.append(mcrx, qargs)
```

Comment:
The implementation of ``QuantumCircuit.mcrx`` is much more efficient, could you change to using that implementation? I.e. something like
```python
qc.mcrx(rule[0].params[0], q_control[:], q_target[0])
```

Reply:
this will add multiple gates to the circuit (mcrx:1 to 5 gates, mcry: 1 to 4 gates depending on num of controls). If you believe that efficiency will be an issue I will adopt your approach.

different implementations results:
mcrx:
![mcrx](https://user-images.githubusercontent.com/28988727/87882595-b3e70680-ca09-11ea-86ac-e3622cba5100.png)

mcry
![mcry](https://user-images.githubusercontent.com/28988727/87882623-e1cc4b00-ca09-11ea-953a-4ce3d6f5dc3a.png)

RXGate.control()
![RXGate control](https://user-images.githubusercontent.com/28988727/87882626-e85ac280-ca09-11ea-844b-f8ce11cf3b37.png)

RYGate.control()
![RYGate control](https://user-images.githubusercontent.com/28988727/87882635-edb80d00-ca09-11ea-8727-4853ae32fd46.png)

Reply:
Actually, we're both not right -- for RX both do the same. Note that the image you showed above, with the double controlled RX, is not what will be run on the hardware, we need to decompose the circuits in terms of a set of basis gates, such as `u1 u3 cx`. If we do that then it turns out they have the same gate count. 
See this snippet:
```python
num_ctrl = 5
q_ctrl = QuantumRegister(num_ctrl)
q_target = QuantumRegister(1)
function = QuantumCircuit(q_ctrl, q_target)
function.mcrx(0.2, q_ctrl[:], q_target[0], None)

gate = QuantumCircuit(q_ctrl, q_target)
gate.append(RXGate(0.2).control(num_ctrl), q_ctrl[:] + q_target[:])
basis = ['u1', 'u3', 'cx']
print('mcrx:', transpile(function, basis_gates=basis).count_ops())
print('RXGate:', transpile(gate, basis_gates=basis).count_ops())
```
prints
```
mcrx: OrderedDict([('cx', 92), ('u3', 62), ('u1', 1)])
RXGate: OrderedDict([('cx', 92), ('u3', 62), ('u1', 1)])
```
For no ancilla qubits, the `RYGate` implementation is even more efficient 👍